Enhance gh issue view to conditionally handle v1 projects being available

jtmcg opened the pull request:
Closes github/cli#490

This has been quite the adventure and has resulted in a decently sized refactor. I learned a ton along the way and am feeling a lot more confident in my ability to work with the rest of the codebase 🙌 Additionally, while doing this I discovered that v2 project support was never added to this command. I've created a follow-up issue, github/cli#544 to track the remaining work.

## Changes

- Extracted all the printing functionality out of `issue/view.go` and into various issue printer interfaces. This allows for better testing, easier to understand code, and the ability to better tease out the requirements of github/cli#490
- Tests were added/expanded to ensure
  - `gh issue view` will only show v1 projects if available
  - `gh issue view --json projectCards` will only show v1 projects if available, showing no data if not available

## Out of scope

This doesn't enable Project v2 support in `gh issue view`. That work is tracked in github/cli#544

## Testing

- Pull down the branch and run the tests: `go test ./...`
- Build the tool locally using `make`
- Run `bin/gh issue view -R github/cli 490`
- Notice that "Projects" doesn't appear in the output (would be under "assignees")

![Screenshot 2024-09-11 at 10 34 31 AM](https://github.com/user-attachments/assets/eb6dbd08-a19a-4a54-bedd-5ec3e3f4df09)

- Run `bin/gh issue view -R github/cli 490 | cat`
- Notice that there are not projects listed

![Screenshot 2024-09-11 at 10 36 04 AM](https://github.com/user-attachments/assets/379aac29-75de-458e-b53f-a6d5b68a3c48)

"Do the hard thing so you can do the easy fix"

jtmcg committed:
Add testing for printHumanIssuePreview around Projects output

To satisfy the first AC in https://github.com/github/cli/issues/490, we
require testing of the output and the feature detector. Seeing as the api
will work for both v1 & v2 endpoints, the first step of is to ensure that
the output isn't showing Projects when no projectCards (a v1 concept) are
included in the issue it is meant to print.

jtmcg committed:
Refactor issue/view/view.go to introduce IssuePrint struct

The entire tty printing of this command was tested through fragile blocks
of strings that made any sort of testing or refactoring difficult, as I
found in the previous commit. I found myself iterating back and forth
with the tests and the code in order to make sure my expected output was
just right to match what the code was doing.

To simplify this, I've introduced the IssuePrint struct, which contains
methods for each of the individual pieces of an issue that will be
printed. This not only makes understanding what the terminal output will
look like easier, but allows us to both test these functions individually
(which I'll do in a forthcoming commit) and mock them for simpler testing
of the larger command.

In this commit, I have not added any functions beyond changing them to
work with the new struct. All tests are green. In addition, there were a
few minor bugfixes that this new format found, such as a nil pointer issue
in the milestone printing section, that have been fixed as well.

jtmcg committed:
Move IssuePrint and tests to separate files

Keeping these separate from the main view.go file will allow for easier
navigation of the code and facilitate better testing practices.

Note: all current tests are essentially integration tests because the
interfaces between the functionality wasn't well defined.

jtmcg committed:
Add test for rawIssuePreview and refactor

Upon adding a test for rawIssuePreview, I realized that the IssuePrint
interface was getting overloaded. Thus, I pulled the humanIssuePreview
and rawIssuePreview functionality out of it, scoping it down to just
formatting the outputs for the human readable issues and renaming it
to IssuePrintFormatter.

Notably, I think getAssigneeList, getLabelList, and getProjectList are all
in the wrong place... It seems to me like they may belong on api.Issue
itself, though that touches something way deeper in the codebase. They
will live on the IssuePrintFormatter interface for now, but I'll probably
want to move them somewhere else before long.

jtmcg committed:
Refactor Issue and Label to handle their own formatted data

This moves getAssigneeList, getProjectList, and getLabelList out of the
IssuePrintFormatter and into either Issue or Label with a small amount of
refactoring to separate functionality that was previously intertwined.

- getAssigneeList -> GetAssigneeListString
- getProjectList  -> GetProjectListString
- getLabelList    -> SortAlphabeticallyIgnoreCase (Labels),
                     getColorizedLabelsList (IssuePrintFormatter)

GetAssigneeListString and GetProjectListString now have tests and can be
called from any Issue

getLabelList saw a few other changes and one design choice. Essentially,
the old getLabelList function was doing two things: sorting the labels
alphabetically, ignoring case, and colorizing the labels if a Color field
was included in the Label. I've separated this functionality into two
places: Labels now handle sorting of themselves with the
SortAlphabeticallyIgnoreCase method on the Labels struct, and
colorization is done by the IssuePrintFormatter.

This ultimately led to me removing colorization from the rawIssuePreview
labels, but can be added back in with a little thought and refactor if we
decide colorization of the rawIssuePreview is desireable.

jtmcg commented on api/queries_issue.go:
```diff
@@ -333,3 +345,41 @@ func (i Issue) Identifier() string {
+	return list
+}
+
+func (i Issue) GetProjectListString() string {
```
Add comments to exported functions/methods

jtmcg committed:
Add tests for the IssuePrintFormatter

This adds tests to cover the methods inside the IssuePrintFormatter.

Note: there is some over-testing in a few of these cases because of how
both Issue and Label are defined. We would need to refactor them to use
interfaces for mocking out this functionality if these tests are to be
pure unit tests. As of now, that seems like a bit too much work for the
scope of this commit.

jtmcg commented on pkg/cmd/issue/view/issuePrintFormatter_test.go:
```diff
@@ -0,0 +1,619 @@
+	}
+}
+
+// Note: this is currently over-testing. We should be able to mock the
+// return from GetAssigneeListString (which has its own tests), but I think
+// that requires a larger refactor of the Issue code to leverage interfaces
+// for mocking.
```
@williammartin I'm struggling a bit to wrap my head around how exactly to leverage an interface with `Label` to allow for mocking of this `GetAssigneeListString` method in these tests... Any insights would be appreciated

williammartin replied:
Riiiiight. I think we're feeling friction here not due to a missing interface but due to an overloading of the `api` domain.

Putting this PR mentally aside for a moment, if I asked you "How does converting an issue's assignees into a string relate to the GitHub API?" I think you would probably say "what?". Or if I said to you, "Is the API responsible for how data is presented?", I think you would probably say "No". These methods `GetAssigneeListString` and `GetProjectListString` and `SortAlphabeticallyIgnoreCase` are presentation logic that this PR is hanging from the API domain type.

If I ask myself "what does the `IssuePrintFormatter` do right now?, I'd say "Well, it takes an issue that is returned from the API, converts that data into a format for presentation, then does some conditional printing of it". Which makes me suspicious that the conversion is a responsibility that should live elsewhere.

My gut feeling here is that we're missing a new domain type in the `view` package that accurately captures what it means to view an issue. One of the fields on that would be `assignees string`, containing the already stringified data.

I would expect a function (naming TBD) like:

```go
func apiIssueToPresentationIssue(api.Issue) presentationIssue {
    // here we can test that assignees are turned to strings correctly
}
```

Then the `IssuePrintFormatter` doesn't have anything to mock, it already has the plain old data it needs.

Let's chat about this more if it's not clear.

jtmcg replied:
Ah, I think I'm following... You're right, that we're overloading the API. This is an excellent idea that will help untangle this further. Thanks!

jtmcg replied:
Alright, I took a stab at this by introducing the `presentationIssue` interface. It did remove the need for mocking 🙌 but I'm sure there's more I'm missing. Let's chat about it in our pairing sesh.

jtmcg commented on pkg/cmd/issue/view/view.go:
```diff
@@ -163,10 +165,14 @@ func findIssue(client *http.Client, baseRepoFn func() (ghrepo.Interface, error),
+	out := IO.Out
+
+	assignees := issue.GetAssigneeListString()
+	// Labels no longer have color in the raw issue preview
```
~~This is a choice I made to clean up the code. Let me know if you agree or disagree. Notably, no tests failed when I made this change 😅~~

Recovered this in a recent commit

williammartin commented on pkg/cmd/issue/view/issuePrintFormatter.go:
```diff
@@ -0,0 +1,132 @@
+package view
```
**Requested Change**

It's conventional in Go (but more importantly, in this codebase) to use underscores in file names so please make that change here and the test file.

jtmcg replied:
Ah, right. Thanks

jtmcg replied:
I think I've done this. Let me know if I missed any

williammartin commented on pkg/cmd/issue/view/view.go:
```diff
@@ -188,140 +194,32 @@ func printRawIssuePreview(out io.Writer, issue *api.Issue) error {
-		text.FuzzyAgo(opts.Now(), issue.CreatedAt),
-		text.Pluralize(issue.Comments.TotalCount, "comment"),
-	)
+func humanIssuePreview(ipf *IssuePrintFormatter, isCommentsPreview bool) error {
```
**Food For Thought**

At first review I think this might benefit from a bit of refactoring around ["Tell Don't Ask"](https://martinfowler.com/bliki/TellDontAsk.html). Instead of giving all this information to a thing and then asking for it back, I suspect it would form better cohesiveness if we had two concrete formatters that we could inject to satisfy an interface, and then called render on those.

I think this would be a good opportunity to pair on. I may be totally wrong, and that would be fine too but it's a nice nuanced concept that will likely appear many times during our time together on the team so spending time on it together seems like it would pay dividends.

jtmcg replied:
Ah, interesting... 🤔 I'll think on that, but pairing does sound like a good idea.

williammartin commented on pkg/cmd/issue/view/issuePrintFormatter.go:
```diff
@@ -0,0 +1,132 @@
+	"github.com/cli/cli/v2/pkg/markdown"
+)
+
+type IssuePrintFormatter struct {
```
**Change Requested**

I don't particularly like the unexported functions because it indicates that you can meaningfully call them separately. However, let's table that discussion in light of my other comment about "Tell, Don't Ask".

What I want to call out here is that this type needs a bit of consistency love. The type `IssuePrintFormatter` is exported and the function `NewIssuePrintFormatter` is exported but none of the methods are exported. So I can see the type and I can construct it but I can't do anything with it (except, work with the `IO` field which also happens to be unexported).

If we continue with this pattern, we should make everything unexported.

jtmcg replied:
TBH, I've got little rhyme or reason to exported vs unexported functions at this point... Would love to discuss some conventions and hear your opinions about them when we pair on this. I'll then take a pass at making these consistent.

williammartin reviewed (commented):
First pass with some focus points. Thanks!

jtmcg committed:
Refactor IssuePrintFormatter with new presentationIssue struct

Based on feedback from @williammartin to my question about using
interfaces for mocking in testing, I've created a new presentationIssue
struct that, when created, handles all the string formatting of an issue
that the IssuePrintFormatter needs. That removed the need for mocking
functions, as the IssuePrintFormatter now just uses any defined
presentationIssue struct.

There's still a few questions I have about this that I expect to discuss
with @williammartin and address before this is completed

jtmcg committed:
Add v2 project support to getProjectListString

williammartin committed:
WIP: Issue Printer Interface

williammartin committed:
WIP: test RawIssuePrinter

jtmcg committed:
Refactor existing IssuePrintFormatter logic into RichIssuePrinter

By pulling the decision logic for what will be printed up to the
NewCmdView function, we were able to tease the functionality previously in
the IssuePrintFormatter out and into two new structs. This completes the
refactoring of the previous "human" readable printing into the
RichIssuePrinter and separates the different data structures into their
own files.

Additionally, I've changed the PresentationIssue being passed around from
a pointer to a copy of the object to avoid any unintended side-effects in
the various methods

jtmcg committed:
Add some comments for thoughts post cleanup

jtmcg commented on pkg/cmd/issue/view/presentation_issue_test.go:
```diff
@@ -0,0 +1,345 @@
+	}
+}
+
+func Test_getProjectListString(t *testing.T) {
```
These are all passing

andyfeller commented on api/queries_comments.go:
```diff
@@ -139,6 +139,8 @@ func (c Comment) Reactions() ReactionGroups {
 	return c.ReactionGroups
 }
 
+// TODO: What is this supposed to do, and how does it relate to
+// formatting comments for PR or Issue view?
 func (c Comment) Status() string {
 	return ""
 }
```
Looking through https://github.com/cli/cli/pull/2757, I believe `Comment` was originally an interface that encompassed issue comments, PR comments, and **PR reviews**.  It is that last one which would have various statuses that would be rendered via: 

https://github.com/cli/cli/blob/192f57ef429b4e89937c87dcbd330520730d4a3a/api/queries_pr_review.go#L115-L117

https://github.com/cli/cli/blob/192f57ef429b4e89937c87dcbd330520730d4a3a/pkg/cmd/pr/shared/comments.go#L166-L184

jtmcg committed:
Move issue label sorting to the presentationIssue

It doesn't make sense that the sorting of labels lives on the api when it
is the presentationIssue that cares about sorting. Thus, I've moved the
sorting functionality to the presentationIssue

jtmcg committed:
Move issue label sorting to the presentationIssue

It doesn't make sense that the sorting of labels lives on the api when it
is the presentationIssue that cares about sorting. Thus, I've moved the
sorting functionality to the presentationIssue

jtmcg committed:
Clean up comments and unused code

jtmcg committed:
Use correct assert in presentation issue tests

andyfeller commented on api/queries_comments.go:
```diff
@@ -139,6 +139,8 @@ func (c Comment) Reactions() ReactionGroups {
 }
 
+// I think that this functionality was moved elsewhere and this has been
+// added so the interface doesn't break
```
I believe this is attempting to satisfy the `Comment` interface which is shared with `PullRequestReview` type:

https://github.com/cli/cli/blob/3cbaffebb6fc621d079592c4d9d117432cfe267a/pkg/cmd/pr/shared/comments.go#L15-L27

https://github.com/cli/cli/blob/3cbaffebb6fc621d079592c4d9d117432cfe267a/api/queries_pr_review.go#L32-L43

It is simply that both are considered comments but no status here whereas PR reviews are considered pending until the review is submitted.

andyfeller commented on pkg/cmd/issue/view/presentation_issue.go:
```diff
@@ -0,0 +1,122 @@
+	return presentationIssue, nil
+}
+
+func getProjectListString(projectCards api.ProjectCards, projectItems api.ProjectItems) string {
```
I find this function name potentially misleading as it makes me think the function returns `[]string`.  I know naming is hard, only thing I can think of is `stringifyProjectItems` 🤷 

_Even that name is more forward thinking once we drop project cards_

jtmcg replied:
Noted. I've updated these accordingly. Let me know if you have any other feedback on the names

andyfeller commented on pkg/cmd/issue/view/presentation_issue_test.go:
```diff
@@ -0,0 +1,398 @@
+				t.Fatal(err)
+			}
+
+			// These are here for development purposes
+			assert.Equal(t, presentationIssue.Title, tc.expect.Title)
+			assert.Equal(t, presentationIssue.Number, tc.expect.Number)
+			assert.Equal(t, presentationIssue.CreatedAt, tc.expect.CreatedAt)
+			assert.Equal(t, presentationIssue.Comments, tc.expect.Comments)
+			assert.Equal(t, presentationIssue.State, tc.expect.State)
+			assert.Equal(t, presentationIssue.Reactions, tc.expect.Reactions)
+			assert.Equal(t, presentationIssue.Author, tc.expect.Author)
+			assert.Equal(t, presentationIssue.AssigneesList, tc.expect.AssigneesList)
+			assert.Equal(t, presentationIssue.LabelsList, tc.expect.LabelsList)
+			assert.Equal(t, presentationIssue.ProjectsList, tc.expect.ProjectsList)
+			assert.Equal(t, presentationIssue.MilestoneTitle, tc.expect.MilestoneTitle)
+			assert.Equal(t, presentationIssue.Body, tc.expect.Body)
+			assert.Equal(t, presentationIssue.URL, tc.expect.URL)
+
+			// This is the actual test
+			assert.Equal(t, presentationIssue, tc.expect)
```
Not suggestion this change, but I always wonder `assert` vs `require`.  In this case here, we would get maximum information if any of the assertions fail for a given test, potentially letting us see if we have multiple problems, yeah?

jtmcg replied:
Oh, thanks for catching this comment. That's right, as that's my understanding of `assert` vs `require` as well. If we used `require`, the test would exit after the first test failure. Say "Title" failed but the rest of the tests were also failing, we wouldn't get the extra context that the issue may be beyond just "Title". I might go fix "Title" in a weird way, as a result, so getting more information here is better.

That said, I don't usually like overloading tests with this many asserts 😅 The comment of "These are here for development purposes" points out that I used these tests to individually tease out the different pieces of this as I refactored. Do you have an opinion on whether or not we should keep them as they are?

andyfeller commented on pkg/cmd/issue/view/rich_issue_printer.go:
```diff
@@ -0,0 +1,146 @@
+func (p *RichIssuePrinter) Print(pi PresentationIssue, repo ghrepo.Interface) error {
+	// I think I'd like to make this easier to understand what the output should look like.
+	// That's probably doable by removing these helpers and just using a formatted string.
+	// I might experiment with that later.
```
Should personal notes like these be moved to the issue, seeing if there is a need for follow up issues?

jtmcg replied:
Another good catch. I've removed this as I'm not sure it's worth following up on

andyfeller commented on pkg/cmd/issue/view/rich_issue_printer_test.go:
```diff
@@ -0,0 +1,405 @@
+				t.Fatal(err)
+			}
+
+			// I can't get these strings to match
+			// assert.Matches(t, stdout.String(), tc.expected)
```
Could you share what happens when this is uncommented?

jtmcg replied:
Ah, thanks for the callout. I discovered I was using the wrong assert after running `go mod tidy` (thanks copilot 🙄). I have switched to the correct one (the `testify` version) and used `assert.Contains()`, which captures this behavior as expected. Thanks for the callout!

andyfeller reviewed (commented):
All in all, tremendous amount of work and appreciate the depth of the tests! ✨ 

Handful of questions regarding personal comments in code, but nothing major to prevent getting this merged.

jtmcg committed:
Clean up comments

There were several comments that were addressed in the PR that I was able
to clean up. Additionally, an old commented-out test was fixed.

jtmcg committed:
Clean up comments

There were several comments that were addressed in the PR that I was able
to clean up. Additionally, an old commented-out test was fixed.

jtmcg committed:
Rename map functions for PresentationIssue

The pattern get<Thing>ListString caused some confusion in the PR about
what the function was doing. I've renamed the helper functions to
stringify<Thing> to clear up that confusion.

andyfeller reviewed (approved):
💪 ✨

williammartin commented on pkg/cmd/issue/view/presentation_issue.go:
```diff
@@ -0,0 +1,122 @@
+	return strings.Join(labelNames, ", ")
+}
+
+func sortAlphabeticallyIgnoreCase(l api.Labels) api.Labels {
```
I think this function signature is a bit confusing given the implementation.

```
sortAlphabeticallyIgnoreCase(l api.Labels) api.Labels
```

This indicates that I give it a labels and it gives me back sorted labels. This is true but it's kind of hiding the fact that the slice of Nodes is being mutated in place. This is also surprising in the way we use it because it's mutating the original `api.Issue` as well.

williammartin replied:
I think this function signature is a bit confusing given the implementation.

```
sortAlphabeticallyIgnoreCase(l api.Labels) api.Labels
```

This indicates that I give it a labels and it gives me back sorted labels. This is true but it's kind of hiding the fact that the slice of Nodes is being mutated in place. This is also surprising in the way we use it because it's mutating the original api.Issue as well.

williammartin commented on pkg/cmd/issue/view/presentation_issue.go:
```diff
@@ -0,0 +1,122 @@
+	Title          string
+	Number         int
+	CreatedAt      time.Time
+	Comments       api.Comments
```
I think it's a bit leaky that we've gone to all this work the separate our `PresentationIssue` from our `api` domain and we're holding on to `api.Comments` in here. I'd need to take some time to investigate a bit more before offering any kind of suggestion though.

williammartin commented on pkg/cmd/issue/view/raw_issue_printer.go:
```diff
@@ -0,0 +1,39 @@
+		return nil
+	}
+
+	// Print empty strings for empty values so the number of metadata lines is consistent when
```
Very nice comment.

williammartin reviewed (commented):
Just a couple of notes. Overall I think this is a big improvement thank you.
